src/libertem/udf/masks.py

```diff
@@ -0,0 +1,148 @@
+        use_torch : bool, optional
+            Use pytorch back-end if available. Default True
+        use_sparse : Union[None, False, True, 'scipy.sparse', 'scipy.sparse.csc', \
+                'sparse.pydata'], optional
+            Which sparse back-end to use.
+            * None (default): Use sparse matrix multiplication if all factory functions return a \
+                sparse mask, otherwise convert all masks to dense matrices and use dense matrix \
+                multiplication
+            * True: Convert all masks to sparse matrices and use default sparse back-end.
+            * False: Convert all masks to dense matrices.
+            * 'scipy.sparse': Use scipy.sparse.csr_matrix (default sparse)
+            * 'scipy.sparse.csc': Use scipy.sparse.csc_matrix
+            * 'sparse.pydata': Use sparse.pydata COO matrix
+        mask_count : int, optional
+            Specify the number of masks if a single factory function is used so that the
+            number of masks can be determined without calling the factory function.
+        mask_dtype : numpy.dtype, optional
+            Specify the dtype of the masks so that mask dtype
+            can be determined without calling the mask factory functions. This can be used to
+            override the mask dtype in the result dtype determination. As an example, setting
+            this to np.float32 means that masks of type float64 will not switch the calculation
+            and result dtype to float64 or complex128.
+        preferred_dtype : numpy.dtype, optional
+            Let :meth:`get_preferred_input_dtype` return the specified type instead of the
+            default `float32`. This can perform the calculation with integer types if both input
+            data and mask data are compatible with this.
+        '''
+        if use_sparse is True:
+            use_sparse = 'scipy.sparse'
+
+        self._mask_container = None
+
+        super().__init__(
+            mask_factories=mask_factories,
+            use_torch=use_torch,
+            use_sparse=use_sparse,
+            mask_count=mask_count,
+            mask_dtype=mask_dtype,
+            preferred_dtype=preferred_dtype,
+            *args, **kwargs
+        )
+
+    def get_preferred_input_dtype(self):
+        ''
+        if self.params.preferred_dtype is None:
+            return super().get_preferred_input_dtype()
+        else:
+            return self.params.preferred_dtype
+
+    def get_mask_dtype(self):
+        if self.params.mask_dtype is None:
+            return self.masks.dtype
+        else:
+            return self.params.mask_dtype
+
+    def get_mask_count(self):
+        if self.params.mask_count is None:
+            return len(self.masks)
+        else:
+            return self.params.mask_count
+
+    @property
+    def masks(self):
+        if self._mask_container is None:
+            self._mask_container = self._make_mask_container()
+        return self._mask_container
```

Comment:
Why do we need the mask container as an instance attribute? Why not only have it in task data? Is it somehow used before `get_task_data` is run? I guess for getting the mask count when creating the result buffers?

I'm thinking ahead a bit, how this will fit together with putting static data into shared memory, but I think for that, the API has to be augmented anyways - `get_task_data` can be used for constructing buffers and other mutable structures, so we'll have to add a new method for static shared data.

Reply:
Yes, I need the mask count and dtype to specify the result buffers. I need this both on the workers and on the central node.

Yes, a static MaskContainer could be very useful.